csharp/src/Apache.Arrow/Schema.cs

```diff
@@ -63,5 +63,68 @@ public int GetFieldIndex(string name, StringComparer comparer = default)
             return _fields.IndexOf(
                 _fields.Single(x => comparer.Compare(x.Name, name) == 0));
         }
+
+        public override bool Equals(object obj)
+        {
+            if (obj == null || !this.GetType().Equals(obj.GetType()))
+            {
+                return false;
+            }
+            return Equals((Schema)obj);
+        }
+
+        public bool Equals(Schema other)
+        {
+            if (this == other)
+            {
+                return true;
+            }
+            var otherFields = other._fields;
+
+            if (this.HasMetadata != other.HasMetadata)
+            {
+                return false;
+            }
+            if (this._fields.Count != otherFields.Count)
+            {
+                return false;
+            }
+            
+            for (int ii = 0; ii < _fields.Count; ii++)
+            {
+                if (!_fields[ii].Equals(otherFields[ii]))
+                {
+                    return false;
+                }
+            }
+
+            if (this.HasMetadata && other.HasMetadata)
+            {
+                return this.Metadata.Keys.Count() == other.Metadata.Keys.Count() &&
+                       this.Metadata.Keys.All(k => other.Metadata.ContainsKey(k) && this.Metadata[k] == other.Metadata[k]) &&
+                       other.Metadata.Keys.All(k => this.Metadata.ContainsKey(k) && other.Metadata[k] == this.Metadata[k]);
+                //return this.Metadata.OrderBy(r => r.Key).SequenceEqual(other.Metadata.OrderBy(r => r.Key));
+            }
+            return true;
+        }
+
+        public override int GetHashCode()
+        {
+            int fieldHashCodes = 0;
```

Comment:
If you want some really good reading on hash codes, check out [https://blogs.msdn.microsoft.com/ericlippert/2011/02/28/guidelines-and-rules-for-gethashcode/](https://blogs.msdn.microsoft.com/ericlippert/2011/02/28/guidelines-and-rules-for-gethashcode/).

There are also numerous articles on stack overflow, like [https://stackoverflow.com/questions/263400/what-is-the-best-algorithm-for-an-overridden-system-object-gethashcode](https://stackoverflow.com/questions/263400/what-is-the-best-algorithm-for-an-overridden-system-object-gethashcode).

The .NET BCL (base class library) finally decided to make it easier to implement GetHashCode by providing some hash APIs:

[https://github.com/dotnet/corefx/issues/14354](https://github.com/dotnet/corefx/issues/14354)

But the issue there is that the new `HashCode` APIs are only available on `.NET Core 2.1`. So we will have to write our own GetHashCode implementations.

Maybe one option is using the [HashCodeCombiner](https://github.com/aspnet/Extensions/blob/master/src/Shared/src/HashCodeCombiner/HashCodeCombiner.cs) from ASP.NET's code? It is already Apache License 2.0, so we shouldn't have any problems with the license.

@wesm - would copying that one file into this repo be an issue?